### PR TITLE
add num_digits function

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed
```python
>>> num_digits(1000, 10)  # no need to len(str(n).lstrip('-'))
4

>>> t=time();[num_digits(-10**(10**i),10) for i in range(6)];time()-t
[2, 11, 101, 1001, 10001, 100001]
0.1290130615234375

>>> t=time();[len(str(-10**(10**i)).lstrip('-')) for i in range(6)];time()-t
[2, 11, 101, 1001, 10001, 100001]
1.5361535549163818
```

#### Other comments

This also physically relocates the int-related functions to intfunc in core so they can be imported without circular imports and reduce the amount of unrelated code in the core files like numbers and power.

When I tried to logically locate them in their original places through import from `intfunc` I got code quality errors -- is this something I should be able to do and *should* do? This PR will break code that relied on importing from those previous locations.

#### Release Notes

* core
  * new file `intfunc.py` was added and is not the place to import new function `num_digits`: returns number of digits in a number in a given base (e.g. `ndigits(100,10) == 3`)
  * files moved to `intfunc` from `power` (isqrt, integer_log and integer_nthroot) and from `numbers` (igcd, igcd2, igcd_lehmer, ilcm, igcdex, and mod_inverse)
* ntheory
  * files moved to `intfunc` from `factor_` (trailing)

### CODE REVIEW

File: sympy/ntheory/digits.py [hunk 1]
```diff
@@ -1,9 +1,51 @@
...
+    if b < 2:
+        raise ValueError('base must be integer greater than 1')
+    n = abs(as_int(n))
+    if n < b:
+        return 1
+    d = 1 + math.floor(math.log10(n)/math.log10(b))
+    s = n//b**d
+    if not s:
+        return d
+    return d + 1  # assert not s//b
```
Comment (haru-44) [inline sympy/ntheory/digits.py:46]:
```suggestion
    d = math.floor(math.log10(n)/math.log10(b))
    b_ = b**d
    while b_ <= n:
        d += 1
        b_ *= b
    return d
```

For example, `ndigits(2**48-1, 2)` would be 49.
In any case, calculation errors are always present, so it is necessary to check each one individually.

File: sympy/ntheory/tests/test_digits.py [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
+    # depending on whether one rounds up or down or uses log or log10,
+    # one or more of these will fail if you don't check for the off-by
+    # one condition
+    assert ndigits(1000, 10) == 4
+    assert ndigits(125, 5) == 4
+    assert ndigits(100, 16) == 2
+    assert ndigits(-1000, 10) == 4
+    assert ndigits(-125, 5) == 4
+    assert ndigits(-100, 16) == 2
+
```
Comment (haru-44) [inline sympy/ntheory/tests/test_digits.py:16]:
```suggestion
    for base in range(2, 100):
        for e in range(1, 100):
            n = base**e
            assert ndigits(n, base) == e + 1
            assert ndigits(n + 1, base) == e + 1
            assert ndigits(n - 1, base) == e

```

Added tests

Comment (smichr) [reply sympy/ntheory/tests/test_digits.py:16]: I had run almost the exact same thing -- likely having been conditioned by reading your PRs ! But when I found these that broke the original algorithm for different rounding/log schemes I took them as a summary. I appreciate the feedback. This is a much more delicate problem than it appears! :-)

**Comment by smichr** [comment]
@haru-44 , do you agree that this is a good function to add to digits?

**Comment by haru-44** [comment]
I don't see any immediate concerns with adding this function to `digits`. 
However, the threshold used in `factor_` is arbitrary, and it should suffice to determine if it's a "large number". For instance, considering a rewrite with a condition like `math.log2(n) > 100` might be worth contemplating.

**Comment by smichr** [comment]
> might be worth contemplating.

Understood. There is only one other place internally that I am looking at for including this (and there, we need the actual length, not an approximation). I'll leave the change in `factor_` there but I don't mind a new PR taking it out.

**Comment by smichr** [comment]
@haru-44 , one more thing. The place that motivated this was `numbers.py`. Does the last commit look like the right way to handle this? I can't import ndigits (at least at top) in numbers.py because of circular imports.

**Comment by haru-44** [comment]
Where to put the `ndigits` implementation depends on the design concept of `sympy`, but I don't think `ndigits` is necessary in last commit. If we want to know the number of digits, we can just call `Integer.length` and there is no need to use `ndigits`.

I don't know the details of sympy, but for example `primetest.py` can't import `trailing` at the top. Can you handle this in the same way?

https://github.com/sympy/sympy/blob/f0605682235df95baaffbf08055c856c6a4e7aa9/sympy/ntheory/primetest.py#L13-L37

**Comment by smichr** [comment]
I'm not sure where to put this. The function is an integer-related function. One should not have to create a SymPy object in order to use it. Perhaps it should be used as a class method `Integer.length(n, b)` and only be referenced that way in the docstring of `digits`. @oscarbenjamin , any ideas?

**Comment by oscarbenjamin** [comment]
I think that this is a reasonable function to be in `ntheory.digits`. If it is needed in `numbers.py` and that causes cyclic import problems then I would put the implementation as `_ndigits` in `numbers.py` and have ntheory import that to define the public `ndigits` function.

On the other hand `gmpy2` has a most likely faster version so perhaps it should go in `sympy.external.ntheory`:
https://gmpy2.readthedocs.io/en/latest/mpz.html#gmpy2.mpz.num_digits

**Comment by oscarbenjamin** [comment]
Perhaps also `num_digits` is a better name?

**Comment by smichr** [comment]
> Can you handle this in the same way?

If you mean put a trailing attribute on Integer, yes that could be done. But again, it feels weird to have to invoke a SymPy function to do an integer calculation.

Perhaps these support functions should go into core.utilities and then they can be imported to the most logical place. I think we did that with the traversal functions and with the misc and interable utilities. Or perhaps just included in core.numbers like `igcd`.

I think I favor just adding `trailing` and `ndigits` to numbers.py.

**Comment by smichr** [comment]
> better name?

Maybe. `nroots` has the connotation of "numerical roots" like `nsolve`. And we already have `count_digits` which gives `multiset(str(n))`. And we have `bitcount`.

Perhaps the guts of the current `ndigits` should just be moved to `integer_log` and then `ndigits` would just be a thin wrapper to return only `integer_log(n, b)[0]`...or don't even have another function. But `num_digits(n)` seems more self-documenting than `integer_log(n, 10)[0]`. I am now +1/4 to adding this.

**Comment by haru-44** [comment]
> If you mean put a trailing attribute on Integer

My point was to put the implementation in `digits.py` and import it in `numbers.py`, but not at the top.


After hearing @oscarbenjamin's idea, I thought it would be a good idea to do the following

`sympy.external.ntheory`

```python
def num_digits(n, base):
    # Concrete Implementation
    pass
```

`sympy.external.gmpy`

```python
from .ntheory import num_digits as python_num_digits

if GROUND_TYPES == 'gmpy':
    def num_digits(n, base)
        if base <= 62:
            return gmpy.num_digits(n, base)
        return python_num_digits(n, base)
else:
    num_digits = python_num_digits
```

`sympy.ntheory.digits`

```python
from sympy.external.gmpy import num_digits

def ndigits(n, base):
    # Need to consider whether to create
    n, base = as_int(n), as_int(base)
    return num_digits(n, base)
```

File: sympy/core/intfunc.py [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+    """
+    b = as_int(base)
+    if b < 2:
+        raise ValueError("base must be integer greater than 1")
+    if b == 2:
+        return bitcount(n)
+    n = abs(as_int(n))
+    if n < b:
+        return 1
+    d = math.floor(math.log10(n) / math.log10(b))
```
Comment (asmeurer) [inline sympy/core/intfunc.py:48]: Why log10? I would expect if anything log2 would be the fastest since integers use a base 2 representation internally.

**Comment by asmeurer** [comment]
How does this compare to `len(str(n))` for base 10 performance-wise? If str is faster maybe we should special case it for b=10 (same for bases 2, 8, and 16 if we can use `bin`, `oct`, and `hex`), although be aware of the `sys.set_int_max_str_digits` limit if we do.

**Comment by asmeurer** [comment]
Oh I didn't notice you already timed it in the OP. What about power of 2 bases though? I would expect those to be much faster to compute since integers are already represented in binary. Indeed, couldn't this just use the `n.bit_count()` method in those cases (or even in every case)?

File: sympy/core/intfunc.py (hunk 3 above)
Comment (smichr) [reply sympy/core/intfunc.py:97]: base 2 special cased now; 10 makes base 10 better and doesn't matter for the other bases

**Comment by smichr** [comment]
> What about power of 2 bases though

now special cased

File: sympy/core/intfunc.py [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+        return 0, False
+    elif n == 0:
+        raise ValueError('n cannot be 0')
+
+    n = as_int(n)
+    b = as_int(b)
+
+    if n < b:
+        return 0, n == 1
+    if b == 2:
```
Comment (asmeurer) [inline sympy/core/intfunc.py:91]: Shouldn't bit_length be used for any power of 2 too?

Comment (smichr) [reply sympy/core/intfunc.py:93]: ok

**Comment by smichr** [comment]
@haru-44 , please take a look again. It's late here and I am not sure that putting it in external makes sense now because it is no longer a pure python function -- it's mostly a thin wrapper to `integer_log`.

File: sympy/core/intfunc.py [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+
+    See Also
+    ========
+    integer_nthroot
+    sympy.ntheory.primetest.is_square
+    sympy.ntheory.factor_.multiplicity
+    sympy.ntheory.factor_.perfect_power
+    """
+    if b < 0:
+        e, t = integer_log(n if n > 0 else -n, -b)
```
Comment (haru-44) [inline sympy/core/intfunc.py:75]:
```suggestion
        e, t = integer_log(abs(n), -b)
```

File: sympy/core/intfunc.py [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        if n < 0:
+            t = t and e % 2 == 1
+        else:
+            t = t and bool(n % 2 if n < 0 else not n % 2)
```
Comment (haru-44) [inline sympy/core/intfunc.py:79]:
```suggestion
            t = t and e % 2 == 0
```
Perhaps a mistake in the modification.

File: sympy/core/intfunc.py [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+        return e, t
+    if b == 1:
+        raise ValueError('base cannot be 1')
```
Comment (haru-44) [inline sympy/core/intfunc.py:82]:
```suggestion
    if b <= 1:
        raise ValueError('base cannot be 0 or 1')
```
Even the original did not take care of the `b == 0` case, but I think it is better to do so.

File: sympy/core/intfunc.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+    if b == 2:
+        e = n.bit_length() - 1
+        return e, trailing(n) == e
+    t = trailing(b)
+    if 2**t == b:
+        e = int(n.bit_length() - 1)//t
+        n_ = 2 << (t*e)
+        return e, n_ == n
+    if b == 1:
+        raise ValueError('b cannot be 1')
```
Comment (haru-44) [inline sympy/core/intfunc.py:102]:
```suggestion
```
Already checked

**Comment by haru-44** [comment]
The fixes to move `trailing` and `integer_log` are not all done yet. The fixes will be extensive and I am not sure if we can pull it off successfully. That is my concern, but I have no objection to the overall plan.

**Comment by smichr** [comment]
@haru-44 as I was working on this it was as I feared -- it is confusing to know how to handle the definitions of functions in external (like isqrt, etc...) And in general (@oscarbenjamin ) it might be separating out "class support, pure Python" functions from the files that define the classes. So `integer_log` doesn't belong in power.py, it belongs somewhere else which will never generate a circular dependency. So every module (perhaps) could have a "utilities" file/folder in which these support functions could go. `intfunc` could be named something else -- that is just a first thought for what was going there (and that name can change), but the more general idea I have not pursued. Any thoughts on this? @oscarbenjamin , I think you did a lot of rearrangement to get fewer import problems so you might have some better ideas of how to deal with this.

**Comment by asmeurer** [comment]
It would be good to have comments or docstrings to that effect in files like core/intfuncs.py. Ideally these functions would all be in ntheory but they are only in the core to avoid circular imports. So, for instance, any additional functions should only go there if they also would have circular import issues otherwise.

**Comment by smichr** [comment]
I am getting an unrelated mamba not installed error.

**Comment by smichr** [comment]
I have no further changes planned unless there are specific requests. Feel free to merge.

**Comment by oscarbenjamin** [comment]
> Ideally these functions would all be in ntheory but they are only in the core to avoid circular imports

I don't agree with this in general. Ideally we would have a layer somewhere of lower-level non-symbolic functions that would not be in either ntheory or core. It is a mistake to think of core as being at the base of the import stack because it is circularly entwined with most of everything. It might be the case that *public* functions for end users should be in ntheory and that the docs for those functions should appear in ntheory. Really though from a code organisation perspective if ntheory is expected to contain anything that depends on the core then it is not the right place for the implementation code for these algorithms (and neither is core).

Some people seem to be objecting to the use of `sympy.external` (because of the name?). I am fine with having something else like `sympy.base` or something but there needs to be a place where these implementations and the logic that manages using gmpy2 etc can go. The circularity is not solved by putting things in core. This also applies to other things like sympy.core.random etc: the core is *not* the base of SymPy's import stack.

Also if we do want to add a `sympy.base` package then its `__init__.py` should not do any recursive imports so that anything from it should be imported with fully qualified imports.

The ntheory module could have been made to be free-standing but the decision to put symbolic functions in there (rather than in `sympy.functions`) prevents that.

**Comment by oscarbenjamin** [comment]
> I am fine with having something else like `sympy.base` or something but there needs to be a place where these implementations and the logic that manages using gmpy2 etc can go

There could be e.g. `sympy.base.ntheory`.

The premise of `sympy.base` would be that it contains anything that does not need to depend on core and would itself be entirely non-circular. Ideally the polys module would not depend on core either but it would depend on `sympy.base.ntheory`. Anything in the core that could be moved to `sympy.base` should be moved there (i.e. anything in the core that does not depend on `Basic`).

**Comment by oscarbenjamin** [comment]
Note that the reason for using `sympy.external` was partly because in some sense some of the things are external (coming from gmpy2) but also just because it is one of the very few places in the codebase that does not recursively import from everywhere else. I just put things there to avoid creating a new top-level package but if the issue with it is the name "external" then let's make a module with a name that people are happy with.

**Comment by haru-44** [comment]
Of the functions we were going to move to `intfunc.py` now, `isqrt`, `integer_nthroot`, and `igcd` are already thin wrappers. Also, `ilcm`, `igcdex`, `num_digits`, `trailing`, and `mod_inverse` can be made thin wrappers by moving their implementations to `external`. I think it is a conservative change to migrate the references to these functions one by one and eventually remove the wrapper functions.
Difficulties should be split.

**Comment by smichr** [comment]
> The circularity is not solved by putting things in core.

I began to see this when I tried to move `perfect_power` and it is much more clear to me that `core != base`. I do like the idea of the name `base` instead of `external` since the latter suggests "external packages" to me instead of "external/independent of the core symbolics". Perhaps it is in there that the shadow structure could be used so there could be `base.ntheory` to help with organization. 

I can do some of the work but I think it would be good to have someone else lay out the game plan.

**Comment by oscarbenjamin** [comment]
I don't mind moving these but let's open an issue with the 1.13 milestone to keep track of what things have been moved where. Anything already moved since 1.12 can be moved somewhere else before 1.13 without breaking any compatibility and without needing to leave placeholder functions/imports.

**Comment by asmeurer** [comment]
Ideally the SymPy submodules should be organized by category. The reason I say these things should go in ntheory is because they are number theory functions. The reality though is that we also have to consider the dependency stack when organizing the submodules, but in principle it would be nice if we didn't, and could just organize things based on where they belong logically. 

This especially applies for imports that end up being used by end-users, but it would also just make the general codebase easier to understand. If someone is looking for the implementation of a function like ndigits, they aren't thinking about whether that is a "core algorithm" or something. They are thinking about what sort of function that is (number theory), and looking in the appropriate place (sympy.ntheory). I personally would never have guessed that num_digits is an important enough function that it gets used in the core (apparently based on the diff here, it only gets used in one small place). 

In general, trying to organize the submodules based on dependencies is bad because it requires people to know about the entire codebase just to guess where something goes. It also is doomed to fail anyway, IMO, because the SymPy submodule dependency graph is inherently cyclical. Submodules should be organized based on higher level logical categories, not things that depend on implementation details.

Unfortunately, I don't think there is really a practical way to avoid thinking about submodules in a dependency sense at least to some degree because of the way modules are actually loaded by Python. But I do at least want to avoid pushing this "disorganization" farther than is absolutely required, which is why I suggested adding a comment to sympy/core/intfuncs.py that only things that _have_ to go there because of circular imports should go there. Everything else should go where it actually makes sense logically, i.e., sympy/ntheory.

**Comment by oscarbenjamin** [comment]
> Ideally the SymPy submodules should be organized by category. The reason I say these things should go in ntheory is because they are number theory functions. The reality though is that we also have to consider the dependency stack when organizing the submodules, but in principle it would be nice if we didn't, and could just organize things based on where they belong logically.

It is possible to do this but it requires being clear about what is in scope for each module and what should go somewhere else. In the case of `ntheory` the decision to add symbolic functions there is where the dependency chain breaks down. Other symbolic functions are in `sympy.functions` and their analogue of the ntheory non-symbolic functions is in mpmath.

**Comment by asmeurer** [comment]
Whether or not a function is symbolic or not is also somewhat of an implementation detail. Right now there are many ntheory functions that aren't symbolic, but they could be. I don't know if it makes sense to move a function from ntheory to functions just because it gains the ability to represent symbolic arguments. And it gets messier if you consider that some ntheory functions aren't exported to the top-level namespace.

**Comment by oscarbenjamin** [comment]
> Whether or not a function is symbolic or not is also somewhat of an implementation detail.

It is definitely not an implementation detail because it directly affects how the function can be used e.g. the symbolic version can do things that the numeric one cannot and they return different types. It also affects speed which perhaps you *might* consider an implementation detail. To me though the difference between "this is intended to be the fastest possible implementation" and "this is something necessarily slower than the fastest possible implementation but also does other things that might not be needed" is relevant when choosing what function to call (I would not call that an implementation detail).

> Right now there are many ntheory functions that aren't symbolic, but they could be. I don't know if it makes sense to move a function from ntheory to functions just because it gains the ability to represent symbolic arguments.

I think firstly we need to distinguish the non-symbolic function from the symbolic one like say `gmp2.factorial` vs `sympy.factorial`. The symbolic factorial function generalises the integer one to complex arguments so it is actually something different. Even purely for integer arguments it is a mistake to try to implement the algorithm for computing the factorial of an explicit integer at the symbolic level when it would always be more efficient to make a pure numeric routine that works with `int/mpz` or just to use an existing more efficient function from an external library like `gmpy2.factorial` or `mpmath.factorial`.

There are several reasons to want to have an efficient integer-only numeric factorial function in say ntheory:

1. Perhaps the other numeric functions in ntheory need it.
2. Perhaps other non-symbolic parts of SymPy outside of ntheory (e.g. polys) would want to use it.
3. Perhaps the symbolic parts of SymPy would want to use it e.g. as part of the implementation of a symbolic `factorial` function (or `gamma` etc).
4. Perhaps users or downstream code might want to use the numeric function directly rather than a necessarily slower symbolic function that always returns symbolic values and might have some other logic that affects its evaluation (e.g. not evaluating `factorial(x)` for enormous arguments or being affects by `evaluate(False)` etc).

Another reason for SymPy to have an efficient numeric only implementation of something like `factorial` is because there are usually going to be competing implementations like a pure Python `sympy/mpmath` one or a sympy/mpmath implementation that can leverage `mpz` or `fmpz` from gmpy2 or python-flint or a more efficient implementation directly from gmpy2 or python-flint. Somewhere there needs to be some logic for selecting the best of these and for defining any wrapper functions to ensure that we can have a single efficient routine that exposes a consistent interface, regardless of which other packages are installed.

Having these numeric functions and centralising the logic for selecting the most efficient is useful both for SymPy and for downstream code and it is easier to arrange this if it is all concentrated in a smaller part of the codebase (e.g. sympy.external or sympy.base) that does not depend on anything else cyclically. The fact that some of these functions are coming from external libraries suggests that they should be near the base of the import stack. It is not completely necessary that everything has to be in a single module somewhere but it is simpler in terms of managing cyclic dependencies to put these things in one part of the codebase that the rest can simply treat as a "library".

So there are good reasons to have a numeric version of `factorial` that is separate from the symbolic version. If there are two versions of `factorial` then only one can be exposed at top-level as `sympy.factorial` but even if that is the symbolic version we would still want to have the numeric version and there should be some way to import it at least for internal use but probably also for downstream code as well.

I would say that ntheory mostly consists of numeric functions. Over time some of those have been replaced by symbolic versions but I think that *replacing* the numeric function is a mistake. It would be better to add a *new* symbolic function in `sympy.functions` that can call the more efficient lower-level numeric routine for `Integer` arguments. We should still keep both versions in the codebase but we can consider which of them should be exposed at top-level as e.g. `sympy.factorial`.

This is all related btw to my comment here about everything not needing to be a Basic subclass: https://github.com/sympy/sympy/pull/25512#issuecomment-1692865797.

**Pull request merged**